### PR TITLE
CosignerSecurityECIES

### PR DESCRIPTION
**Pull request by agouin**
Second phase of #162 

Adds ECIES implementation of `CosignerSecurity`

Deprecates RSA CosignerSecurity

### CODE REVIEW

**Review by mark-rushakoff — APPROVED** [review]
Looks pretty straightforward. I think there are some opportunities to spread key generation work across available cores; alternatively, tests could use cached, deterministic keys. Either way is helpful to ensure tests stay fast going forward.

File: cmd/horcrux/cmd/root.go [hunk 1]
```diff
@@ -24,7 +24,15 @@ func rootCmd() *cobra.Command {
 	cmd.AddCommand(startCmd())
 	cmd.AddCommand(addressCmd())
 	cmd.AddCommand(createCosignerEd25519ShardsCmd())
-	cmd.AddCommand(createCosignerRSAShardsCmd())
+	cmd.AddCommand(createCosignerECIESShardsCmd())
+
+	rsaCmd := createCosignerRSAShardsCmd()
+	rsaCmd.Deprecated = `
+ECIES is recommended for cosigner-to-cosigner encryption since it is much faster.
```
[the remark this answers is not in the document]
Comment (agouin) [reply cmd/horcrux/cmd/root.go:31]: Your made up example is very close. We are using the secp256k1 curve with a private key bit size of 256 bits. This is considered to be as secure as RSA 3072. I will add a comment.

https://sectigostore.com/blog/ecdsa-vs-rsa-everything-you-need-to-know/

File: signer/cosigner_security_ecies_test.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	require.NoError(t, err)
+
+	decryptedPub, decryptedShare, err := s2.DecryptAndVerify(1, nonce.PubKey, nonce.Share, nonce.Signature)
+	require.NoError(t, err)
+
+	require.Equal(t, mockPub, decryptedPub)
+	require.Equal(t, mockShare, decryptedShare)
+
+	_, _, err = s3.DecryptAndVerify(1, nonce.PubKey, nonce.Share, nonce.Signature)
+	require.ErrorContains(t, err, "cannot decrypt ciphertext: cipher: message authentication failed")
```
[the remark this answers is not in the document]
Comment (agouin) [reply signer/cosigner_security_ecies_test.go]: good call, refactored.

File: signer/cosigner_security_ecies.go [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+	digestBytes, err := cometjson.Marshal(digestMsg)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	digest := sha256.Sum256(digestBytes)
+	pubKey, ok := c.eciesPubKeys[id]
+	if !ok {
+		return nil, nil, fmt.Errorf("unknown cosigner: %d", id)
+	}
```
[the remark this answers is not in the document]
Comment (agouin) [reply signer/cosigner_security_ecies.go]: no risk, good catch

File: cmd/horcrux/cmd/threshold.go [hunk 4]
```diff
@@ -24,9 +25,15 @@ func NewThresholdValidator(
...
-	if err != nil {
-		return nil, nil, err
+	var security signer.CosignerSecurity
+	var eciesErr error
+	security, eciesErr = config.CosignerSecurityECIES()
+	if eciesErr != nil {
+		var rsaErr error
+		security, rsaErr = config.CosignerSecurityRSA()
+		if rsaErr != nil {
+			return nil, nil, errors.Join(eciesErr, rsaErr)
```
[the remark this answers is not in the document]
Comment (agouin) [reply cmd/horcrux/cmd/threshold.go]: good add

**Pull request merged**